src/main/kotlin/org/elm/ide/intentions/MakeDecoderIntention.kt

```diff
@@ -0,0 +1,184 @@
+package org.elm.ide.intentions
+
+import org.elm.lang.core.imports.ImportAdder.Candidate
+import org.elm.lang.core.lookup.ElmLookup
+import org.elm.lang.core.psi.ElmFile
+import org.elm.lang.core.psi.elements.ElmTypeDeclaration
+import org.elm.lang.core.types.*
+
+class MakeDecoderIntention : AnnotationBasedGeneratorIntention() {
+    override fun getText() = "Generate Decoder"
+
+    override fun getRootIfApplicable(annotationTy: Ty): Ty? {
+        if (annotationTy !is TyUnion) return null
+        if (annotationTy.module != "Json.Decode" || annotationTy.name != "Decoder") return null
+        return annotationTy.parameters.singleOrNull()
+    }
+
+    override fun generator(context: Context): TyFunctionGenerator {
+        return DecoderGenerator(context.file, context.ty, context.name)
+    }
+}
+
+private class DecoderGenerator(
+        file: ElmFile,
+        root: Ty,
+        functionName: String
+) : TyFunctionGenerator(file, root) {
+    /** true if we need to import Json.Decode.Pipeline */
+    private var usedPipeline = false
+
+    override val code by lazy {
+        // run gen on the root to kick off generation
+        val genBody = gen(root)
+        val rootFunc = funcsByTy[root]
+        funcsByTy.remove(root)
+        val body = rootFunc?.body ?: "    $genBody"
+
+        buildString {
+            append("\n$functionName =\n")
+            append(body)
+
+            for (f in funcsByTy.values) {
+                append("\n\n\n")
+                append("${f.name} : ${qual("Decoder")} ${f.qualifier}${f.paramTy.renderedText(false, false)}\n")
+                append("${f.name} =\n")
+                append(f.body)
+            }
+        }
+    }
+
+    override fun calculateImports(): List<Candidate> {
+        return when {
+            usedPipeline -> super.calculateImports() + listOf(
+                    Candidate(
```

Comment:
Probably should rename `Candidate` since it is now being used in a different way.